File: 547-number-of-provinces/number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        city_connections = {}
        city_count = len(isConnected)

        for city in range(city_count):
            city_connections[city] = []

            for connection in range(city_count):
                if city == connection: continue
                if isConnected[city][connection] == 1:
                    city_connections[city].append(connection)

        providence_count = 0
        interconnected_cities = set()

        for city in city_connections:
            if city in interconnected_cities: continue
            interconnected_cities.add(city)
            stack = [*city_connections[city]]
            while stack:
                curr_city = stack.pop()
                if curr_city in interconnected_cities: continue
                stack.extend(city_connections[curr_city])
                interconnected_cities.add(curr_city)

            providence_count += 1

        return providence_count

File: 547-number-of-provinces/number-of-provinces.java
class Solution {
    HashMap<Integer, int[]> cityConnections;
    HashSet<Integer> seenCities = new HashSet<>();
    int providenceCount = 0;

    public int findCircleNum(int[][] isConnected) {
        // This is another graph problem
        // Given a matrix of edge connections, these seem to be undirected edges? IE connection can be traversed from one node to another and back
        // Need to find the number of distinct trees in the graph, ie groups of nodes
        // Can create a hashMap of cityConnections
        // Can then traverse through each city in cityConnections and go through all the potential connections, only counting providences on the outter traversal
        this.cityConnections = createHashMap(isConnected);
        traverseProvidences(this.cityConnections);
        return this.providenceCount;
    }

    public HashMap<Integer, int[]> createHashMap(int[][] edgeMatrix) {
        HashMap<Integer, int[]> hashMap = new HashMap<>();

        for(int i = 0; i < edgeMatrix.length; i ++) {
            hashMap.put(i, edgeMatrix[i]);
        }

        return hashMap;
    }

    public void traverseProvidences(HashMap<Integer, int[]> cityConnections) {
        for(int city : cityConnections.keySet()) {
            if(!this.seenCities.contains(city)) {
                this.seenCities.add(city);
                this.providenceCount += 1;
                int[] connections = cityConnections.get(city);
                _traverseConnections(connections, cityConnections);
            }
        }
    }

    public void _traverseConnections( int[] connections, HashMap<Integer, int[]> cityConnections) {
        for(int i = 0; i < connections.length; i ++) {
            if(connections[i] == 1 && !this.seenCities.contains(i)) {
                this.seenCities.add(i);
                int[] newConnections = cityConnections.get(i);
                _traverseConnections(newConnections, cityConnections);
            }
        }
    }
}